### ISSUE TITLE
casting from NULL inside ANY() prevents column from being used for optimization

### ISSUE THREAD

**Issue by arctica** [cockroachdb/cockroach#42562]
Cockroach Version: 19.2

I've hit a strange performance bug recently while using ANY($1) in a NULL parameter.

A query of the form `SELECT x FROM t WHERE a = ANY($1)` where $1 is NULL will run extremely slowly because it can't use the column a in an index scan.

Example outputs:

```
root@:26257/db> select count(1) from timeseries where a = any(cast(null as int[]));
  count
+-------+
      0
(1 row)

Time: 31.452ms

root@:26257/db> select count(1) from timeseries where a = any(cast('{}' as int[]));
  count
+-------+
      0
(1 row)

Time: 1.321ms
```

Note the runtime difference. Below also outputs for explain:

```
root@:26257/db> explain select count(1) from timeseries where a = any(cast(null as int[]));
       tree      |    field    |            description
+----------------+-------------+------------------------------------+
                 | distributed | true
                 | vectorized  | false
  group          |             |
   │             | aggregate 0 | count(column18)
   │             | scalar      |
   └── render    |             |
        └── scan |             |
                 | table       | timeseries@timeseries_a_b_idx
                 | spans       | ALL
                 | filter      | a = ANY CAST(NULL AS INT8[])
(10 rows)

Time: 2.512ms

root@:26257/db> explain select count(1) from timeseries where a = any(cast('{}' as int[]));
     tree     |    field    |   description
+-------------+-------------+-----------------+
              | distributed | true
              | vectorized  | false
  group       |             |
   │          | aggregate 0 | count(column18)
   │          | scalar      |
   └── norows |             |
(6 rows)

Time: 1.399ms
```

PS: I'm not sure if I explained the problem properly in the title of the issue, feel free to edit anything as you see fit.

PPS: It would be nice if EXPLAIN anyways showed the table/index it used to see that no rows are matching instead of just "norows".

**Comment by jordanlewis** [comment]
Hmm, for the second one, I think the optimizer can infer that there will be no matches since the array is empty - that's why it says `norows`. What if you put some integers in the array?

**Comment by arctica** [comment]
> Hmm, for the second one, I think the optimizer can infer that there will be no matches since the array is empty - that's why it says `norows`. What if you put some integers in the array?

Right, so the optimizer even without looking at the table can see that the condition can never be true if the array is empty no matter the table contents so doesn't even look at it. I guess that makes sense.

To your second question: if I put integers in the array then the query properly uses the right index to scan and no "filter".

**Comment by ricardocrdb** [comment]
Hey @arctica 

The first query is also taking a longer time as it needs to do a full table scan, indicated by the `scan spans ALL` section of the `EXPLAIN` output, along with the information provided by @jordanlewis .

Let me know if there are any other questions, or if we can close this issue.

**Comment by arctica** [comment]
@ricardocrdb I don't think closing the ticket is the correct thing because the bug in question hasn't been addressed yet at all. @jordanlewis gave an explanation for the "norows" output of the second query. The first query is still behaving wrongly. As you correctly saw it is doing a scan over all spans but in fact it should behave like the second one and realize there are no rows matching. At least that's my understanding. To me it looks like the query planner can't handle the NULL in this case properly.

**Comment by jordanlewis** [comment]
Yeah, I agree that there's probably an issue here. Can you show `SHOW CREATE TABLE` on the table in question?

**Comment by arctica** [comment]
I've removed a lot of columns because there are quite a few and I don't think they are involved but here it goes:

```
CREATE TABLE timeseries (
ts TIMESTAMP NOT NULL,
a INT8 NOT NULL,
b INT8 NOT NULL,
val1 DECIMAL NULL,
val2 DECIMAL NULL,
CONSTRAINT "primary" PRIMARY KEY (ts ASC, a ASC, b ASC),
INDEX timeseries_a_ts_idx (a ASC, ts ASC),
FAMILY "primary" (ts, a, b, val1, val2)
);

insert into timeseries values(now(), 1, 1, 1, 1);

explain select count(1) from timeseries where a = any(cast('{}' as int[]));

explain select count(1) from timeseries where a = any(cast(null as int[]));
```

I verified this minimal example shows the behaviour in question.

**Comment by jordanlewis** [comment]
Oh, I see now. I agree that this is wrong. `a = any(null)` for any value of `a` should return `null`, and `null` in any `where` clause should always evaluate to false.

But is this what you expect, that your query would return no rows?

I think we have a bug with casting null to an int array - here's a quick repro:

```
root@127.0.0.1:64360/defaultdb> create table a (a int, b int, index(a,b));
CREATE TABLE

Time: 2.34ms

root@127.0.0.1:64360/defaultdb> explain select * from a where a=any(null);
   tree  |    field    | description
+--------+-------------+-------------+
         | distributed | false
         | vectorized  | false
  norows |             |
(3 rows)

Time: 379µs

root@127.0.0.1:64360/defaultdb> explain select * from a where a=any(null::int[]);
  tree |    field    |         description
+------+-------------+------------------------------+
       | distributed | true
       | vectorized  | false
  scan |             |
       | table       | a@primary
       | spans       | ALL
       | filter      | a = ANY CAST(NULL AS INT8[])
(6 rows)

Time: 545µs
```

I wonder if that's what you expect though? Were you looking for a query that returns true if `a is null`? Because the query you've chosen isn't going to do that.

As far as the bug, @RaduBerinde could you take a look at this? The filter in my example should be provably false, so it should turn into `norows` like the same query without the cast.

**Comment by RaduBerinde** [comment]
Just to clarify because the word "wrong" has been used. There are no correctness issues here, just performance (not getting the best execution plan), correct?

**Comment by jordanlewis** [comment]
Yes, that's right.

**Comment by RaduBerinde** [comment]
Ok. The fix should be fairly simple, just a new simple folding rule. Will have a change out soon.

**Comment by RaduBerinde** [comment]
@arctica By default, our fixes go to the next major version (20.1). Is there a strong desire to have this fix backported onto a 19.2.x release (likely 19.2.2)?

**Comment by arctica** [comment]
@jordanlewis nope I never expected any rows in the results for the case when NULL is supplied as input. So the result is fine, just was an unexpected performance degradation.

@RaduBerinde that was a quick fix, nice. For me personally there's no strong desire to have a fix "right now" as I was able to easily sidestep the problem by making the application supply an empty array which is then converted to CAST('{}' AS int[]) when before it would just send a NULL instead. So all good. And since noone else reported this problem before I guess it affects roughly ANY(NULL) people :)
